### ISSUE TITLE
Registering logger/trainer inside nested PL models

### ISSUE THREAD

**Issue by CompRhys** [Lightning-AI/lightning#13146]
### Discussed in https://github.com/PyTorchLightning/pytorch-lightning/discussions/13144

<div type='discussions-op-text'>

<sup>Originally posted by **CompRhys** May 24, 2022</sup>
`You are trying to 'self.log()' but the 'self.trainer' reference is not registered on the model yet. This is most likely because the model hasn't been passed to the 'Trainer'`

I have a PL module that I am now incorporating inside another PL module as part of a more complicated setup. Inside the sub module there are still various things that I would like to log but as it's not being trained directly instead via a trainer on the overall module the self.trainer is not set. Is it okay to just pass self.trainer of the overall model as an additional kwarg to the sub model or could this have unforeseen consequences?</div>

Looking at the code the log command checks if `trainer` is `None`,  looking at the implications of this it seems as though there's no real way to resolve this issue of nested logging due to many things depending on the state of the trainer. I am using a module to wrap another module to train an ensemble of small models where I can do early stopping based on the validation loss of the ensemble. I would also like to log the individual losses of the ensemble members but can't see a way to handle this in PL.

cc @carmocca @edward-io @ananthsub @rohitgr7 @kamil-kaczmarek @Raalsky @Blaizzy @justusschock @awaelchli @borda @ninginthecloud @jjenniferdai

**Comment by carmocca** [comment]
Honestly, this is not a usecase we've ever explored.

You can try setting the `self.trainer` reference recursively for all submodules. You might encounter other complications though.

Why exactly do the sub-modules have to be `LightningModules`? Why not have them be `nn.Modules` that process  inputs and return outputs that you `self.log` in the main `LightningModule`?

**Comment by CompRhys** [comment]
I want to essentially test the trade off in resource use between ensemble and single model given fixed retraining compute budget. The single model is setup as a PL module and then was exploring based on discussion this idea of wrapping it in a PL module to make the ensemble. Despite the benefits of ensembles I haven't yet found a nice way to package/train them.

**Comment by carmocca** [comment]
Is your code public? It would be interesting to check it out.

**Comment by CompRhys** [comment]
The Deep Ensemble logic mimics this code https://github.com/PyTorchLightning/pytorch-lightning/discussions/8505#discussioncomment-1036824 but my training step calls the training steps of the PL modules inside the ensemble 

```python
    def training_step(self, batch, batch_idx, optimizer_idx):
        model = self.ensemble[optimizer_idx]
        loss, outputs = model.training_step(batch, batch_idx, optimizer_idx, return_preds=True)
        ...
```

I could ask to share the code if the above snippet isn't enough but not sure what response I would get

**Comment by CompRhys** [comment]
I gave this a quick try but didn't have much success but I didn't give any thought to what the above code was actually doing yet. I will spend some time over the weekend looking at it and report back proper in the next week or so.

More generally I guess the question is that is there any aim to add ensemble functionality to the trainer object? It would probably require introduction of some new callbacks to handle reduction over the ensemble for the type of whole ensemble logging I am interested in - this would need to be non-blocking so as to not kill performance due to waiting but the engineering around handling ensembles is non trivial (imo) and would be nice to see efforts to abstract away or come up with better/best practices.

**Comment by carmocca** [comment]
We wouldn't add in-house ensemble support directly to the Trainer, but we are always open to improvements to our internals that allow deeper levels of extensibility.
